### ISSUE TITLE
🚀 Feature: Allow a way for data migrations to be rolled back when downgrading backstage

### ISSUE THREAD

**Issue by mhaley37** [backstage/backstage#22439]
### 🔖 Feature description

Hi, 
The `backstage-cli` should be updated to allow data migrations to be rolled back -- ideally, this would be an implicit process ( i.e. if you try to downgrade backstage from version `B` to version `A`, then data migrations in `B` but not `A` should be rolled back).   However if this is too ambiguous, it would still be useful to add commands to `backstage-cli` to allow this in a more explicit process.

### 🎤 Context

🧵 [Discord Thread](https://discord.com/channels/687207715902193673/1195419527362658344) where it was suggested for me to submit a new issue.

Recently we deployed a new version of Backstage to a pre-production environment which included multiple data migrations ( e.g. [20230428155633_sessions.js](https://github.com/backstage/backstage/blob/v1.18.4/plugins/auth-backend/migrations/20230428155633_sessions.js) ).  This process went well and the migrations were applied to the database correctly.

Later, we decided to rollback to the previous version.  however we encountered a startup failure on the backend with the following error:

```
Backend failed to start up Error: The migration directory is corrupt, the following files are missing: 20230428155633_sessions.js
```

I understand _why_ this is happening -- when we updated to the new version, new rows were added to `knex_migrations` for the new migrations, and those files do not exist in the previous version.

This made rollback difficult.   We ended up solving this by doing the following for each missing migration:
1. Connect to our DB and manually run logic in the `down` method of the migration ( e.g. [20230428155633_sessions.js down method](https://github.com/backstage/backstage/blob/v1.18.4/plugins/auth-backend/migrations/20230428155633_sessions.js)
2. Remove the corresponding row from the `knex_migrations` table
3. Retry

This worked but isn't ideal.  Moreover, there isn't an easy way to know _which_ migrations are part of a new release.  The error at startup only listed a single migration failure, and after solving this it would fail on the next missing migration.  It would be really useful if the [Backstage Upgrade Helper](https://backstage.github.io/upgrade-helper/) could include data migrations between the two version as well, or a `backstage-cli` command to do something similar.

For context, we are running in a Kubernetes environment so it isn't uncommon for us to create pods with newer versions of backstage for testing purposes in our lower environments.

### ✌️ Possible Implementation

The migration files all look to include an exported `down` method for the express purpose of undoing a migration, so `backstage-cli` could hook into this the same way it must for the `up` method when it _applies_ the migration.   Also, when comparing two versions, a diff of the migrations can be found simply by looking at the `migrations` path of each plugin and comparing the two.   While this can be done manually, it is extremely tedious and error prone.

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

None

**Comment by vinzscam** [comment]
Hey @mhaley37, thanks for flagging this 🙏
Cross linking #6366 as the issue is very much related.

**Comment by webark** [comment]
Should the migration down run happen when a service connects to the DB (like applying them happens now)? Or should it need to be something which is required to be manually ran?

**Comment by mhaley37** [comment]
> Should the migration down run happen when a service connects to the DB (like applying them happens now)? Or should it need to be something which is required to be manually ran?

It would be great if this was baked into the backend startup as part of the db migrations checks ( theorized flow):

1. Rollback to version 1 ( from 2)
2. Backend starts up at version 1
3. Backend finds migrations in the `knex_migrations` table that aren't present in that version
4. (new) Backend identifies that these migrations are "from the future" ( Maybe by checking the date part of the migration filename, and comparing that against the date version 1 was cut?)
5. (new) Run the `down` method for all migrations identified as "from the future", remove from `knex_migrations`
6. (new) Start the migration process again ( so that normal migration checks still take place)

I'd probably settle just for a `cli` entry point, as I could at least clean this up quicker than now, which there is no real good method of going about it.

**Comment by benjdlambert** [comment]
I think that there's a little issue that we don't have the migration code for future releases, so we can't rollback on older versions in step 5.

I think that we're gonna get some time later this year to take a look at this unless there's some other contributions that come in to help get this moving :pray:

**Comment by mhaley37** [comment]
> I think that there's a little issue that we don't have the migration code for future releases, so we can't rollback on older versions in step 5.

Yeah, this would have to reach out to the repo or something similar to determine what migrations are part of <future release>, so it wouldn't work in air-gapped deployments, but otherwise this is not unlike `backstage-cli versions:bump` no?

**Comment by margueritepd** [comment]
Semi off-topic but in the hopes of helping others who need to roll back a migration: here is how I did it.

```
$ node_modules/.bin/knex  migrate:currentVersion --connection "postgresql://$POSTGRES_USER:$POSTGRES_PASSWORD@$POSTGRES_HOST/backstage_plugin_app" --client pg
Using environment: production
Current Version: 20240113144027

$ node_modules/.bin/knex  --connection 
"postgresql://$POSTGRES_USER:$POSTGRES_PASSWORD@$POSTGRES_HOST/backstage_plugin_app" --client pg --migrations-directory node_modules/@backstage/plugin-app-backend/migrations/ migrate:status
Using environment: production
Found 2 Completed Migration file/files.
20211229105307_init.js
20240113144027_assets-namespace.js
No Pending Migration files Found.

$ node_modules/.bin/knex  --connection "postgresql://$POSTGRES_USER:$POSTGRES_PASSWORD@$POSTGRES_HOST/backstage_plugin_app" --client pg --migrations-directory node_modules/@backstage/plugin-app-backend/migrations/ migrate:down 20240113144027_assets-namespace.js
Using environment: production
Batch 2 rolled back the following migrations:
20240113144027_assets-namespace.js

$ node_modules/.bin/knex  --connection "postgresql://$POSTGRES_USER:$POSTGRES_PASSWORD@$POSTGRES_HOST/backstage_plugin_app" --client pg --migrations-directory node_modules/@backstage/plugin-app-backend/migrations/ migrate:status
Using environment: production
Found 1 Completed Migration file/files.
20211229105307_init.js
Found 1 Pending Migration file/files.
20240113144027_assets-namespace.js
```

**Comment by jhaals** [comment]
Probably too big for contribfest. Happy to look at an RFC _before_ starting work as this is a large task.

**Comment by webark** [comment]
One thing that could be nice is if the service just didn't start if there were migrations, and didn't do migrations on startup. And migrations had to be handled completely separately. 🤔

**Comment by angeliski** [comment]
Amazing @margueritepd , saved my life S2

@Rugvip Do you see any problem to include that in the docs? Before #6366 should be really helpful

**Comment by MihaiTabara** [comment]
> Amazing [@margueritepd](https://github.com/margueritepd) , saved my life S2
> 
> [@Rugvip](https://github.com/Rugvip) Do you see any problem to include that in the docs? Before [#6366](https://github.com/backstage/backstage/issues/6366) should be really helpful

Discussed this in today's triaging meeting. Let's bake this in our documentation as it indeed may help others hitting this issue. 
@angeliski would you be open to help document this?

**Comment by angeliski** [comment]
Sure @MihaiTabara 

Do you have any suggestion about where we could put that info? Maybe https://backstage.io/docs/backend-system/core-services/database?

**Comment by awanlin** [comment]
There's no obvious clear cut place for this, I might add it here : https://backstage.io/docs/tutorials/configuring-plugin-databases or as a stand alone tutorial.

 Then cross link it in a few places: like the deployment section, the place you mentioned, and here: https://backstage.io/docs/getting-started/config/database

**Comment by webark** [comment]
Still want this!

**Comment by lexhuismans-wk** [comment]
When developing plugins and creating migrations we encounter similar issues. Especially tricky if a problematic deployment ends up in production and rollback is not straight forward.

From my understanding it will currently always require manual intervention, making updates to the DB. Which we generally try to avoid.

**Comment by webark** [comment]
not stale. This still makes upgrades tricky as it's much harder to automatically rollback if there is an issue.
